pkg/syndesis/action/upgrade.go

```diff
@@ -0,0 +1,194 @@
+package action
+
+import (
+	"errors"
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+	"github.com/sirupsen/logrus"
+	"github.com/syndesisio/syndesis-operator/pkg/apis/syndesis/v1alpha1"
+	syndesistemplate "github.com/syndesisio/syndesis-operator/pkg/syndesis/template"
+	"github.com/syndesisio/syndesis-operator/pkg/syndesis/version"
+	"github.com/syndesisio/syndesis-operator/pkg/util"
+	"k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"time"
+)
+
+const (
+	UpgradePodServiceAccountName = "syndesis-operator"
+)
+
+// Upgrades Syndesis to the version supported by this operator using the upgrade template.
+type Upgrade struct {
+	operatorVersion	string
+}
+
+func (a *Upgrade) CanExecute(syndesis *v1alpha1.Syndesis) bool {
+	return syndesisInstallationStatusIs(syndesis, v1alpha1.SyndesisInstallationStatusUpgrading)
+}
+
+func (a *Upgrade) Execute(syndesis *v1alpha1.Syndesis) error {
+	if a.operatorVersion == "" {
+		operatorVersion, err := version.GetSyndesisVersionFromOperatorTemplate()
+		if err != nil {
+			return err
+		}
+		a.operatorVersion = operatorVersion
+	}
+
+	namespaceVersion, err := version.GetSyndesisVersionFromNamespace(syndesis.Namespace)
+	if err != nil {
+		return err
+	}
+	targetVersion := a.operatorVersion
+
+	resources, err := a.getUpgradeResources(syndesis)
+	if err != nil {
+		return err
+	}
+
+	templateUpgradePod, err := a.findUpgradePod(resources)
+	if err != nil {
+		return err
+	}
+
+	upgradePod, err := a.getUpgradePodFromNamespace(templateUpgradePod, syndesis)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	if syndesis.Status.ForceUpgrade || k8serrors.IsNotFound(err) {
+		// Upgrade pod not found or upgrade forced
+
+		if namespaceVersion != targetVersion {
+			logrus.Info("Upgrading syndesis resource ", syndesis.Name, " from version ", namespaceVersion, " to ", targetVersion)
+
+			// Set the correct service account for the upgrade pod
+			templateUpgradePod.Spec.ServiceAccountName = UpgradePodServiceAccountName
```

Comment:
Shouldn't we set the service account already in the template itself? Well, it probably doesn't harm to set it here, too, but we would have one less hardcoded value in the go code.

Reply:
Yes, there are other changes in the template respect to the original template that is present in Syndesis and this can be also changed there. The issue will be then how we keep the operator template in sync with the one in the syndesis repo... but this is already a issue we have to solve...

Reply:
I think when we manage to move over syndesis-operator to syndesis then we can directly pick it up from a relative directory. Of course we then would still think about how to manage the productized templates, which currently live in a different directory. The idea here is, that we pick up the operator from syndesis operator and build it in fuse-ignite-install with the productised templates and give it a productised name.